### PR TITLE
Fix #41388 by preserving protocol and port when generating routes

### LINKED ISSUE #41388: Active Storage - Disk Service does not keep the host protocol in development for direct uploads
### Steps to reproduce

Hi,

I was currently debugging an upload error that was only happening in development and found out that it had to do with direct uploads and the server running with SSL.

I have a file input with the option `direct_upload: true` and I'm running the server, with a self-signed SSL certificate, using the command:

```
bundle exec rails s -b 'ssl://localhost:3000?key=localhost-key.pem&cert=localhost.pem'
```

While debugging, I found out that
```ruby
url_helpers.update_rails_disk_service_url(verified_token_with_expiration, host: current_host)
```
from [this line](https://github.com/rails/rails/blob/291a3d2ef29a3842d1156ada7526f4ee60dd2b59/activestorage/lib/active_storage/service/disk_service.rb#L89) is the one having the issue.

### Expected behavior

```ruby
url_helpers.update_rails_disk_service_url(verified_token_with_expiration, host: current_host)
```
should return a URL with the same protocol that `current_host` has.

### Actual behavior

```ruby
url_helpers.update_rails_disk_service_url(verified_token_with_expiration, host: current_host)
```
is returning a URL with `http` as the protocol, not keeping the `https` that `current_host` has, which in my case was `https://localhost:3000`

### System configuration
**Rails version**: 6.1.1

**Ruby version**: 2.6.6

### Note

This is related to the closed issue: https://github.com/rails/rails/issues/40248

### PR DESCRIPTION
**Pull request by santib**
### Summary

Fixes #41388

### CODE REVIEW

File: activestorage/test/service/disk_service_test.rb [hunk 1]
```diff
@@ -12,6 +12,22 @@ class ActiveStorage::Service::DiskServiceTest < ActiveSupport::TestCase
...
+  test "url_for_direct_upload" do
+    original_url_options = Rails.application.routes.default_url_options.dup
+    Rails.application.routes.default_url_options.merge!(protocol: "http", host: "test.example.com", port: 3001)
+
+    key      = SecureRandom.base58(24)
+    data     = "Something else entirely!"
+    checksum = Digest::MD5.base64digest(data)
+
+    begin
+      assert_match(/^https:\/\/example.com\/rails\/active_storage\/disk\/.*$/,
```
Comment (santib) [inline activestorage/test/service/disk_service_test.rb:24]: this test was inspired in the test below, what I don't understand is why the port isn't being preserved in both tests 🤔

**Comment by santib** [comment]
@rafaelfranca would you mind checking this simple fix? I'd really appreciate having it included in a 6.1.x version because at our project we use SSL in both dev and test environments.

**Comment by santib** [comment]
@kaspth would you mind taking a look at it? Notice that the same solution was already in place for `generate_url`, so this change was pretty much following it. I'm needing it because I use SSL in dev env and we are using direct uploads.

**Comment by kaspth** [comment]
Cool! Though let's do our due diligence and clean the campsite while we're here:

```ruby
# For url_for_direct_upload
url_helpers.update_rails_disk_service_url(verified_token_with_expiration, default_url_options).tap do |generated_url|
  payload[:url] = generated_url
end

# For generate_url
url_helpers.rails_disk_service_url(verified_key_with_expiration, filename: filename, **default_url_options)

# Replace current_host_uri with this. No need for memoization, imo.
def default_url_options
  URI.parse(ActiveStorage::Current.host).then do |host_uri|
    { protocol: host_uri.scheme, host: host_uri.host, port: host_uri.port }
  end
end
```

**Comment by santib** [comment]
@kaspth That definitively looks better! TIL about `.then` 😄 

The only thing that I'm not sure about is the name `default_url_options` since it has a special meaning, like if it's defined in a controller it will automatically affect all generated routes which is not the case here.

**Comment by kaspth** [comment]
True, I think it's ok though. The passing is manual, but the naming still carries the same meaning: these are the default url options.

What's `Rails.application.routes.default_url_options` set to in a real Rails app, maybe we could do `Rails.application.routes.default_url_options.merge(host: ActiveStorage::Current.host)` instead?

**Comment by santib** [comment]
> What's `Rails.application.routes.default_url_options` set to in a real Rails app, maybe we could do `Rails.application.routes.default_url_options.merge(host: ActiveStorage::Current.host)` instead?

hmm but what if the `Rails.application.routes.default_url_options` defines a different port/protocol than the value set in `ActiveStorage::Current.host` ? For example in the tests `ActiveStorage::Current.host` is set to `https://example.com` which has a protocol. And in `app/controllers/concerns/active_storage/set_current.rb` is set to a full url `request.base_url`. Do you think we should just ignore protocol/port set in `ActiveStorage::Current.host` ? It's a bit weird that it's called `host` but stores a full url though 🤔

**Comment by georgeclaghorn** [comment]
Not liking the way we make `DiskService` parse the details it actually needs from a value that exists specifically for this purpose. How about replacing `ActiveStorage::Current.host` with `ActiveStorage::Current.url_options` or similar?

File: activestorage/app/models/active_storage/current.rb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 # frozen_string_literal: true
 
 class ActiveStorage::Current < ActiveSupport::CurrentAttributes #:nodoc:
-  attribute :host
+  attribute :url_options
```
Comment (kaspth) [inline activestorage/app/models/active_storage/current.rb:4]: I think what @georgeclaghorn was indicating is that we also do the parsing in this class, like this:

```suggestion
  attribute :url_options

  def set_url_options_from(host)
    self.url_options = URI.parse(host).then do |uri|
      { protocol: uri.scheme, host: uri.host, port: uri.port }
    end
  end
end
```

File: activestorage/lib/active_storage/service/disk_service.rb [hunk 3]
```diff
@@ -164,8 +155,10 @@ def url_helpers
         @url_helpers ||= Rails.application.routes.url_helpers
       end
 
-      def current_host
-        ActiveStorage::Current.host
+      def default_url_options
+        URI.parse(ActiveStorage::Current.url_options).then do |host_uri|
+          { protocol: host_uri.scheme, host: host_uri.host, port: host_uri.port }
+        end
```
Comment (kaspth) [inline activestorage/lib/active_storage/service/disk_service.rb:161]:
```suggestion
      def url_options
        ActiveStorage::Current.url_options
```

File: activestorage/app/models/active_storage/current.rb (hunk 2 above)
Comment (georgeclaghorn) [reply activestorage/app/models/active_storage/current.rb:4]: I don’t think we need to do any parsing. We can set the right URL options based on `request` in `SetCurrent`.

Comment (kaspth) [reply activestorage/app/models/active_storage/current.rb:4]: Ah lol, of course 👍

**Comment by zzak** [comment]
@santib Could you give this a rebase please? :bow:

**Comment by santib** [comment]
> @santib Could you give this a rebase please? 🙇

@zzak Done!

**Comment by zzak** [comment]
@santib Thanks! I'm wondering since we are removing `ActiveStorage::Current.host` we should at least add a changelog? Are there other docs we should consider?

As a possible middle-ground, would it make sense to say merge the `Current.host` value with the default options so apps currently dependent on this don't break?

**Comment by santib** [comment]
> @santib Thanks! I'm wondering since we are removing `ActiveStorage::Current.host` we should at least add a changelog? Are there other docs we should consider?
> 
> As a possible middle-ground, would it make sense to say merge the `Current.host` value with the default options so apps currently dependent on this don't break?

Sure I can add a CHANGELOG. Regarding adding more code to
>merge the Current.host value with the default options so apps currently dependent on this don't break?

I'm not sure it's worth it. This change just affects the DiskService which AFAIK should only be used in dev/testing envs, right? If that's not true then yeah I guess we might want to add a migration path.. maybe adding an `alias` `host` -> `url_options` and a deprection asking people to change their usages of `Current.host` to `Current.url_options`? WDYT?

**Comment by zzak** [comment]
@santib This looks much better IMO, but let's see if others have some feedback. :pray:

**Comment by santib** [comment]
Hi @kaspth, would you mind checking this PR out whenever you have some time? Thanks 🙌

**Review by kaspth — APPROVED** [review]
Sweet! Needs a rebase and commits squashed then we're good to go.

**Comment by santib** [comment]
> Sweet! Needs a rebase and commits squashed then we're good to go.

Done! 😄

**Comment by kaspth** [comment]
Awesome, thanks so much 😄🙌

**Pull request merged**